include/envoy/network/connection.h

```diff
@@ -62,7 +63,10 @@ class ConnectionCallbacks {
  */
 enum class ConnectionCloseType {
   FlushWrite, // Flush pending write data before raising ConnectionEvent::LocalClose
-  NoFlush     // Do not flush any pending data and immediately raise ConnectionEvent::LocalClose
+  NoFlush,    // Do not flush any pending data and immediately raise ConnectionEvent::LocalClose
+  HalfClose,  // Send a half-close after flushing any pending data. If we have already received
```

Comment:
I hadn't looked at the h2 style stuff.  That's another option.  But it might be weird because we'll always end up writing either (data, false), or (empty, true) because receiving the half-close is a connection-event, not included with onData().  But I suppose the half-close-rx could be included with onData() if we wanted.  

Thinking about this, that actually may be more correct due to the following case:

1) some data is read, a network filter in the chain before tcp_proxy gets onData() and returns StopIteration
2) connection receives half-close, sends onEvent(half-close) to all filters
3) tcp_proxy sends half-close to peer filter
4) the other network filter unblocks the data, and now tcp_proxy gets some data that it can't proxy because it has already half-closed.

So would it make sense then to have the half-close signaling always go with the data (both tx and rx)?

Reply:
Yeah, we definitely need to make sure everything is drained before the half close.   I suspect your way of doing things is cleaner from an API perspective but I don't know if there's a way to solve close getting ahead of data with it.  I'm totally open to either if you find a way :-)  

If you go the data/fin route you can totally "combine" data with the FIN for efficiency.